spynnaker/pyNN/models/neuron/population_machine_vertex.py

```diff
@@ -384,6 +397,227 @@ def get_binary_file_name(self):
+            key=routing_info.get_first_key_from_pre_vertex(
+                self, constants.SPIKE_PARTITION_ID),
+            spec=spec,
+            region_id=constants.POPULATION_BASED_REGIONS.NEURON_PARAMS.value)
+
+        # close spec
+        spec.end_specification()
+
+    @overrides(AbstractRewritesDataSpecification.reload_required)
+    def reload_required(self):
+        return self.__change_requires_neuron_parameters_reload
+
+    @overrides(AbstractRewritesDataSpecification.set_reload_required)
+    def set_reload_required(self, new_value):
+        self.__change_requires_neuron_parameters_reload = new_value
+
+    def _reserve_memory_regions(self, spec, machine_graph, n_key_map):
+        """ Reserve the DSG data regions.
+
+        :param ~.DataSpecificationGenerator spec:
+            the spec to write the DSG region to
+        :param ~.MachineGraph machine_graph: machine graph
+        :param n_key_map: n key map
+        :return: None
+        """
+        spec.comment("\nReserving memory space for data regions:\n\n")
+
+        # Reserve memory:
+        spec.reserve_memory_region(
+            region=POPULATION_BASED_REGIONS.SYSTEM.value,
+            size=common_constants.SIMULATION_N_BYTES,
+            label='System')
+
+        self._reserve_neuron_params_data_region(spec)
+
+        spec.reserve_memory_region(
+            region=POPULATION_BASED_REGIONS.NEURON_RECORDING.value,
+            size=self._app_vertex.neuron_recorder.get_static_sdram_usage(
+                self.vertex_slice),
+            label="neuron recording")
+
+        profile_utils.reserve_profile_region(
+            spec, POPULATION_BASED_REGIONS.PROFILING.value,
+            self._app_vertex.n_profile_samples)
+
+        # reserve bit field region
+        bit_field_utilities.reserve_bit_field_regions(
+            spec, machine_graph, n_key_map, self,
+            POPULATION_BASED_REGIONS.BIT_FIELD_BUILDER.value,
+            POPULATION_BASED_REGIONS.BIT_FIELD_FILTER.value,
+            POPULATION_BASED_REGIONS.BIT_FIELD_KEY_MAP.value)
+
+        self.reserve_provenance_data_region(spec)
+
+    @staticmethod
+    def neuron_region_sdram_address(placement, transceiver):
+        return helpful_functions.locate_memory_region_for_placement(
+                placement, POPULATION_BASED_REGIONS.NEURON_PARAMS.value,
+                transceiver)
+
+    def _reserve_neuron_params_data_region(self, spec):
+        """ Reserve the neuron parameter data region.
+
+        :param ~data_specification.DataSpecificationGenerator spec:
+            the spec to write the DSG region to
+        :return: None
+        """
+        params_size = self._app_vertex.get_sdram_usage_for_neuron_params(
+            self.vertex_slice)
+        spec.reserve_memory_region(
+            region=POPULATION_BASED_REGIONS.NEURON_PARAMS.value,
+            size=params_size, label='NeuronParams')
+
+    def _write_neuron_parameters(self, spec, key, region_id):
```

Comment:
Why have a key and are params when
key=routing_info.get_first_key_from_pre_vertex(
                self, constants.SPIKE_PARTITION_ID),
region_id=constants.POPULATION_BASED_REGIONS.NEURON_PARAMS.value

Reply:
not true. you might give the neurons a subset of the keys, at any aribitary offset. we dont know. and if we're allowing it to be overriden then the key is the base entry point

Reply:
Interesting as key = "SPIKE" what would the subset be?

Reply:
your making assumptions here. i might want the neuron keys to be in the middle and have dop at the beginning. who the hell knows. thus leaving the leaver open is what ive done. given the splitter is the one whom is making the amchine vertex. not the app vertex. so the app vertex cant assumer the key is what your stating

Reply:
Discussed with @alan-stokes . Decided to leave flexible for later splitters